src/data_storage/validators.py

```diff
@@ -0,0 +1,59 @@
+from itertools import chain
+
+from utilities.exceptions import (
+    ExistsCircleValidationError,
+    RootDoesNotExistValidationError,
+    UnconnectedNodeValidationError,
+    UnrelatedReferenceValidationError,
+)
+
+
+def deep_first_search(graph, current_vertex, color=None):
+    if color is None:
+        color = {}
+
+    # If we have reference to None or node without options it means branch has end
+    if current_vertex is None or not graph[current_vertex]:
+        return False
+    color[current_vertex] = "grey"
+
+    circle_result = []
+    for node in graph[current_vertex]:
+        node_color = color.get(node)
+        if node_color is None:  # it is expected that all nodes have white color
+            circle_result.append(deep_first_search(graph, node, color))
```

Comment:
Recursion is good to implement algo fast, but from memory perspective it's better to use stack instead.